flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/UnreachableCoalesceArgumentsRemoveRule.java

```diff
@@ -0,0 +1,197 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.planner.plan.utils.FlinkRexUtil;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+
+import java.util.List;
+
+/**
+ * Removes unreachable {@link BuiltInFunctionDefinitions#COALESCE} arguments.
+ *
+ * <p>An unreachable COALESCE argument is defined as any argument after the first argument in the
+ * argument list with a non-null type.
+ */
+@Internal
+public class UnreachableCoalesceArgumentsRemoveRule
+        extends RelRule<UnreachableCoalesceArgumentsRemoveRule.Config> {
+
+    public static final RelOptRule PROJECT_INSTANCE =
+            Config.EMPTY.as(Config.class).withProject().toRule();
+    public static final RelOptRule FILTER_INSTANCE =
+            Config.EMPTY.as(Config.class).withFilter().toRule();
+    public static final RelOptRule JOIN_INSTANCE =
+            Config.EMPTY.as(Config.class).withJoin().toRule();
+    public static final RelOptRule CALC_INSTANCE =
+            Config.EMPTY.as(Config.class).withCalc().toRule();
+
+    private static final UnreachableCoalesceArgumentsRemoveRexShuttle REX_SHUTTLE_INSTANCE =
+            new UnreachableCoalesceArgumentsRemoveRexShuttle();
+
+    public UnreachableCoalesceArgumentsRemoveRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final RelNode relNode = call.rel(0);
+        call.transformTo(relNode.accept(REX_SHUTTLE_INSTANCE));
+    }
+
+    private static class UnreachableCoalesceArgumentsRemoveRexShuttle extends RexShuttle {
+
+        @Override
+        public RexNode visitCall(RexCall call) {
+            call = (RexCall) super.visitCall(call);
+
+            // Not a coalesce invocation, skip it
+            if (!call.getOperator()
+                    .getName()
+                    .equals(BuiltInFunctionDefinitions.COALESCE.getName())) {
+                return call;
+            }
+
+            for (int firstNullableIndex = 0;
+                    firstNullableIndex < call.operands.size();
+                    firstNullableIndex++) {
+                RexNode argument = call.operands.get(firstNullableIndex);
+                // Check if it's non null
```

Comment:
nit: this comment only restates what the code does, and the code is already very expressive, so I would remove it.